examples/nlp/gpt/conf/gpt_spin.yaml

```diff
@@ -0,0 +1,179 @@
+defaults:
+  - optional tp_overlap@model.ub_tp_comm_overlap_cfg:
+
+trainer:
+  num_nodes: 8
+  devices: 8
+  accelerator: gpu
+  precision: bf16-mixed
+  
+  # spin specific args
+  spin:
+    max_iterations: 3   # the number of iterations (rounds) for which the SPIN "game" is played. This is T in the original paper
+    max_epochs: 1       # the number of epochs to use per iteration. In the paper, they do 3 iterations with 2 epochs per iteration
+    max_steps: -1
+    val_check_interval: 0.1
+    save_interval: 100
+    limit_steps_per_epoch: -1
```

Comment:
Actually, thinking about it a bit more I think this won't work because the sampler won't account for it and thus its internal epoch count will be incorrect, right?
Maybe we can remove this option for now and add it back later? (I also want to make some changes to the sampler so it may be best to wait before modifying it to support this)

Reply:
Good point, the sampler will have an incorrect internal epoch count given the way I have written it. We would need to have an accessor method on the sampler itself which allows us to internally change its `total_samples` however, I'm not sure if that would work because the core dataloader would still be wrong in terms of `__len__`. Will need to think on this some more.

In the meantime, this can be accomplished manually by just changing the input dataset size, but that is annoying to need to do.

Reply:
I have changed the parameter name and default value to be in line with PTL. It works fine as is, but we'd need to make some changes to the Sampler to have it behave as PTL intended

Reply:
Ok I'll look into fixing it when I do these sampler changes.

But I think this is going to crash with `limit_train_batches=null`:
```python
if (limit_train_batches := self.cfg.get("limit_train_batches", -1)) > 0:
```
=> should be instead
```python
if (limit_train_batches := self.cfg.get("limit_train_batches")) is not None:
```

Reply:
Doesn't making it null in the config yaml mean that when you call .get on self.cfg, it returns a None which causes it to populate the -1 as the default value?

Reply:
No, `.get()` on a dict only returns the default value if the key is missing:
```
In [14]: {"foo": None}.get("foo", -1) is None
Out[14]: True
```

Reply:
Okay, but we also need to check that the value is > 0, because someone can just change the null to -1 and it would pass your if check. Can we do something like:

```
if (limit_train_batches := self.cfg.get("limit_train_batches")) is not None and limit_train_batches > 0:
            self.num_steps_per_epoch = min(self.num_steps_per_epoch, limit_train_batches)
```

Reply:
From what I can tell from the PTL docs `limit_train_batches < 0` isn't allowed, that being said someone may mistakenly use `-1` instead of `None` so maybe you can just add below the `if`:
```python
assert limit_train_batches >= 0, "`limit_train_batches` must be either a non-negative integer or `None`"
```

Reply:
It would be nice to keep it to a one-liner, what do you think of what I have checked in just now?

Reply:
I have a slight preference for crashing to stick to PTL's API, but it's not a huge deal.

That being said though I took a quick look at how PTL is using it and it looks like a value of 0 is allowed to skip training so it may be better to use `>=` instead of `>`: https://github.com/Lightning-AI/lightning/blob/34a34a0754d612965b158f1f588336507554fa16/src/lightning/pytorch/loops/fit_loop.py#L194

Reply:
This is fixed now